### PR TITLE
[16.0][MIG] sale_order_type

### PR DESCRIPTION
**Pull request by augusto-weiss**

### CODE REVIEW

**Review by marylla — APPROVED** [review]
I made a functional review and it works as expected. :+1: 

@augusto-weiss Can you check the failing tests?

**Comment by augusto-weiss** [comment]
Thanks @marylla for your review... I've already fixed the tests

**Comment by marylla** [comment]
@pedrobaeza Hi Pedro, can you review as well and merge? :)

**Comment by pedrobaeza** [comment]
/ocabot migration sale_order_type

Sorry, you have to enroll another reviewer.

**Comment by victoralmau** [comment]
Please, cherry-pick https://github.com/OCA/sale-workflow/pull/2260

**Comment by augusto-weiss** [comment]
Done @victoralmau !

**Review by nicomacr — APPROVED** [review]
LGTM

**Comment by tate11** [comment]
odoo.api.create: The model odoo.addons.sale_order_type.models.sale is not overriding the create method in batch

gives a warning message

File: sale_order_type/models/account_move.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        string="Sale Type",
+        compute="_compute_sale_type_id",
+        store=True,
+        readonly=True,
+        states={"draft": [("readonly", False)]},
+        ondelete="restrict",
+        copy=True,
+    )
+
+    @api.depends("partner_id", "company_id")
```
Comment (rousseldenis) [inline sale_order_type/models/account_move.py:22]:
```suggestion
    @api.depends("partner_id", "company_id")
    @api.depends_context("default_move_type", "active_model", "company")
```

File: sale_order_type/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    def _compute_sale_type_id(self):
+        # If create invoice from sale order, sale type will not computed.
+        if not self.env.context.get("default_move_type", False) or self.env.context.get(
+            "active_model", False
+        ) in ["sale.order", "sale.advance.payment.inv"]:
+            return
+        self.sale_type_id = self.env["sale.order.type"]
+        for record in self:
+            if record.move_type not in ["out_invoice", "out_refund"]:
+                record.sale_type_id = self.env["sale.order.type"]
```
Comment (rousseldenis) [inline sale_order_type/models/account_move.py]:
```suggestion
                record.sale_type_id = sale_type
```

See here above

File: sale_order_type/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    )
+
+    @api.depends("partner_id", "company_id")
+    def _compute_sale_type_id(self):
+        # If create invoice from sale order, sale type will not computed.
+        if not self.env.context.get("default_move_type", False) or self.env.context.get(
+            "active_model", False
+        ) in ["sale.order", "sale.advance.payment.inv"]:
+            return
+        self.sale_type_id = self.env["sale.order.type"]
```
Comment (rousseldenis) [inline sale_order_type/models/account_move.py]:
```suggestion
        sale_type = self.env["sale.order.type"].browse()
        self.sale_type_id = sale_type
```

File: sale_order_type/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                    or record.partner_id.commercial_partner_id.with_company(
+                        record.company_id
+                    ).sale_type
+                )
+                if sale_type:
+                    record.sale_type_id = sale_type
+
+    @api.onchange("sale_type_id")
+    def onchange_sale_type_id(self):
+        # TODO: To be changed to computed stored readonly=False if possible in v14?
```
Comment (rousseldenis) [inline sale_order_type/models/account_move.py]: Indeed, as invoice_payment_term_id and journal_id are computed fields (readonly=False), you should override the compute methods instead.

File: sale_order_type/models/account_move.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            "active_model", False
+        ) in ["sale.order", "sale.advance.payment.inv"]:
+            return
+        self.sale_type_id = self.env["sale.order.type"]
+        for record in self:
+            if record.move_type not in ["out_invoice", "out_refund"]:
+                record.sale_type_id = self.env["sale.order.type"]
+                continue
+            else:
+                record.sale_type_id = record.sale_type_id
```
Comment (rousseldenis) [inline sale_order_type/models/account_move.py]: I'm not sure of this. Maybe :

```suggestion
                record.sale_type_id = record._origin.sale_type_id
```

is better

File: sale_order_type/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            ],
+            order="company_id",
+            limit=1,
+        )
+
+    @api.depends("partner_id", "company_id")
+    def _compute_sale_type_id(self):
+        for record in self:
+            if not record.partner_id:
+                record.type_id = self.env["sale.order.type"].search(
```
Comment (rousseldenis) [inline sale_order_type/models/sale.py:56]: Please add a test for this case.

File: sale_order_type/models/sale.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                ("code", "=", "sale.order"),
+                "|",
+                ("company_id", "=", force_company),
+                ("company_id", "=", False),
+            ],
+            order="company_id",
+            limit=1,
+        )
+
+    @api.depends("partner_id", "company_id")
```
Comment (rousseldenis) [inline sale_order_type/models/sale.py:51]:
```suggestion
    @api.depends("partner_id", "company_id")
    @api.depends_context("partner_id", "company_id", "company")
```

File: sale_order_type/models/sale.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                )
+            if vals:
+                order.update(vals)
+            # Order line values
+            line_vals = {}
+            line_vals.update({"route_id": order_type.route_id.id})
+            order.order_line.update(line_vals)
+
+    @api.model
+    def create(self, vals):
```
Comment (rousseldenis) [inline sale_order_type/models/sale.py]: As @tate11 said, change this to multi implementation

File: sale_order_type/models/sale.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        if vals.get("name", _("New")) == _("New") and vals.get("type_id"):
+            sale_type = self.env["sale.order.type"].browse(vals["type_id"])
+            if sale_type.sequence_id:
+                vals["name"] = sale_type.sequence_id.next_by_id(
+                    sequence_date=vals.get("date_order")
+                )
+        return super(SaleOrder, self).create(vals)
```
Comment (rousseldenis) [inline sale_order_type/models/sale.py]:
```suggestion
        return super().create(vals)
```

File: sale_order_type/models/sale_order_type.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        return self.env["sale.order"].fields_get(allfields=["picking_policy"])[
+            "picking_policy"
+        ]["selection"]
+
+    @api.model
+    def default_picking_policy(self):
+        default_dict = self.env["sale.order"].default_get(["picking_policy"])
+        return default_dict.get("picking_policy")
+
+    name = fields.Char(required=True, translate=True)
```
Comment (rousseldenis) [inline sale_order_type/models/sale_order_type.py:11]: Please put fields on top and change default fields attributes to lambda.

**Review by rousseldenis — CHANGES_REQUESTED** [review]

**Comment by augusto-weiss** [comment]
Hi @rousseldenis, thanks for your comments... I implemented some changes in a new commit so we could check it.

**Comment by rousseldenis** [comment]
@augusto-weiss Last comment not attended

**Comment by augusto-weiss** [comment]
Hi @rousseldenis, i didn´t do that because methods (_get_domain_sequence_id and _get_selection_picking_policy) are used in the definition of the fields

**Comment by rousseldenis** [comment]
> Hi @rousseldenis, i didn´t do that because methods (_get_domain_sequence_id and _get_selection_picking_policy) are used in the definition of the fields

@augusto-weiss You can change the attribute to lambda and then, it's ok

**Comment by chandni299** [comment]
@pedrobaeza @rousseldenis Can we merge this PR?

File: sale_order_type/models/account_move.py (hunk 1 above)
Comment (rousseldenis) [reply sale_order_type/models/account_move.py:22]: @augusto-weiss This one is not attended

**Review by rousseldenis — APPROVED** [review]

**Comment by rousseldenis** [comment]
@augusto-weiss Could you squash your commits ?

**Comment by augusto-weiss** [comment]
Done @rousseldenis

**Pull request merged**